File: game-of-life/typescript-fpts-game-of-life/src/GameInteraction.ts
import {puts} from "./Console";
import {alive, dead, Matrix, showMatrix} from "./Data";
import {pipe} from "fp-ts/pipeable";
import {map, Task} from "fp-ts/Task";

export const welcome: Task<void> = puts("Welcome to the Conway's Game of Life!")

export const print: (matrix: Matrix) => Task<Matrix> = (matrix: Matrix) => pipe(
    puts(showMatrix.show(matrix)),
    map(() => matrix)
)

export const readInitialMatrix: Task<Matrix> = () => Promise.resolve([
    [dead(), alive(), dead()],
    [dead(), alive(), dead()],
    [dead(), alive(), dead()]
])

File: game-of-life/swift-bow-game-of-life/swift-bow-game-of-life/GameInteraction.swift
//
//  GameInteraction.swift
//  swift-bow-game-of-life
//
//  Created by Fabrizio Duroni on 01.10.20.
//

import Foundation
import BowEffects

func welcome() -> IO<Never, Void> {
    return puts(message: "Welcome to the Tennis Game!")
}

func readInitialMatrix() -> IO<Never, Matrix> {
    return IO.invoke {
        parse(input: try! String(contentsOfFile: Bundle.main.path(forResource: "6x6-toad", ofType: "csv")!))
    }
}

func parse(input: String) -> Matrix {
    return input
        .split(separator: "\n")
        .map { line in
            return line
                .split(separator: ",")
                .map { element in
                    switch(element) {
                    case "Dead": return .Dead
                    default: return .Alive
                    }
                }
        }
        
}

func print(matrix: Matrix) -> IO<Never, Matrix> {
    return puts(message: show(matrix: matrix)).map { matrix }^
}

